app/models/attachment.rb

```diff
@@ -1,9 +1,30 @@
 class Attachment < ActiveRecord::Base
 
+  mount_uploader :asset, AssetUploader
+
   belongs_to :parent, polymorphic: true
 
+  validates :asset, presence: true
   validates :parent, presence: true
-  validates :asset, presence: true,
-                    uniqueness: { scope: [:parent_type, :parent_id] }
+  validate :unique_asset
+
+  protected
+
+  def remove_asset!
+    super unless Attachment.where(asset: asset.file.try(:filename)).exists?
+  end
+
+  def remove_previously_stored_asset
```

Comment:
It's used by carrierwave (I'm redefining their callback to not remove images unless all references are gone).